### PR TITLE
bpo-32424: Synchronize xml.etree.ElementTree.Element copy methods.

### PR DESCRIPTION
**Pull request by GPHemsley**
The Python implementation had `Element.copy()` while the C
implementation had `Element.__copy__()` and `Element.__deepcopy__()`.

This synchronizes the two implementations and deprecates
`elem.copy()` in favor of `copy.copy(elem)`.

https://bugs.python.org/issue32424

### CODE REVIEW

File: Lib/test/test_xml_etree.py [hunk 1]
```diff
@@ -115,6 +120,75 @@ def test_all(self):
...
+        self.assertIs(element_foo2.text, self.element_foo.text)
+        self.assertIs(element_foo2.tail, self.element_foo.tail)
+
+        self.assertEqual(element_foo2.tag, self.element_foo.tag)
+        self.assertEqual(element_foo2.attrib, self.element_foo.attrib)
+        self.assertEqual(element_foo2.text, self.element_foo.text)
+        self.assertEqual(element_foo2.tail, self.element_foo.tail)
+
+        for (child1, child2) in zip(self.element_foo, element_foo2):
+            self.assertIs(child1, child2)
```
Comment (pganssle) [inline Lib/test/test_xml_etree.py]: I dunno if this is a problem with the code coverage software or what, but it doesn't seem like this statement is getting hit by the coverage checker. Is it possible that one of these two has length zero?

I'll note that I personally prefer to use `itertools.zip_longest` for tests like this, since `zip` will just silently fail if one of the two iterators is a truncated version of the other - e.g. `all(x == y for x, y in zip((1, 2, 3), (1, 2, 3, 4, 5)))`.

Comment (GPHemsley) [reply Lib/test/test_xml_etree.py]: The reason the statement wasn't getting hit was because I forgot to actually check it.

And while the lists are expected to be of identical length, I agree that failing silently is not a good idea.

I've created a new commit that addresses these two issues, plus others.

File: Lib/test/test_xml_etree.py [hunk 2]
```diff
@@ -102,6 +103,11 @@ def newtest(*args, **kwargs):
     return decorator
 
 
+#-----------------------------------------------------------------------------
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:113]: This comment is redundant.

File: Lib/test/test_xml_etree.py [hunk 3]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
         support.check__all__(self, ET, names, blacklist=("HTML_EMPTY",))
 
 
+class ElementTree_Element_UnitTest(unittest.TestCase):
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:131]: The name of this class is not in line with other names.

File: Lib/test/test_xml_etree.py [hunk 4]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
...
+
+    def test___init__(self):
+        tag = "foo"
+        attrib = { "zix": "wyp" }
+
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:136]: Too much empty lines. Remove unneeded empty lines.

File: Lib/test/test_xml_etree.py [hunk 5]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
...
+        element_foo = ET.Element(tag, attrib)
+
+        with self.subTest("traits of an element"):
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:139]: This is not needed. It doesn't match the style of other tests.

File: Lib/test/test_xml_etree.py [hunk 6]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
...
+            self.assertIsInstance(element_foo, ET.Element)
+            self.assertIn("tag", dir(element_foo))
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:141]: This is already tested in test_interface.

File: Lib/test/test_xml_etree.py [hunk 7]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
...
+            self.assertIn("attrib", dir(element_foo))
+            self.assertIn("text", dir(element_foo))
+            self.assertIn("tail", dir(element_foo))
+
+        with self.subTest("string attributes have expected values"):
+            self.assertEqual(element_foo.tag, tag)
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:147]: This is already tested in test_set_attribute.

File: Lib/test/test_xml_etree.py [hunk 8]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
...
+            self.assertIsNone(element_foo.text)
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:148]: This may be an implementation detail.

File: Lib/test/test_xml_etree.py [hunk 9]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
...
+            self.assertIsNone(element_foo.tail)
+
+        with self.subTest("attrib is a copy"):
+            self.assertIsNot(element_foo.attrib, attrib)
+            self.assertEqual(element_foo.attrib, attrib)
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:153]: This is already tested in test_attrib.

File: Lib/test/test_xml_etree.py [hunk 10]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
...
+
+        with self.subTest("attrib isn't linked"):
+            attrib["bar"] = "baz"
+
+            self.assertIsNot(element_foo.attrib, attrib)
+            self.assertNotEqual(element_foo.attrib, attrib)
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:159]: This is already tested in test_attrib.

File: Lib/test/test_xml_etree.py [hunk 11]
```diff
@@ -115,6 +121,143 @@ def test_all(self):
...
+            self.assertIsNot(element_foo2.attrib, element_foo.attrib)
+            self.assertEqual(element_foo2.attrib, element_foo.attrib)
+
+        with self.subTest("attrib isn't linked"):
+            element_foo.attrib["bar"] = "baz"
+
+            self.assertIsNot(element_foo2.attrib, element_foo.attrib)
+            self.assertNotEqual(element_foo2.attrib, element_foo.attrib)
+
+    def test___copy__(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:195]: There is an existing test for copying: test_copy.

File: Lib/test/test_xml_etree.py [hunk 12]
```diff
@@ -1786,36 +1929,38 @@ def test_cyclic_gc(self):
...
-        e2 = ET.Element('e2')
-        e3 = ET.Element('e3')
-        e1.append(e2)
-        e2.append(e2)
-        e3.append(e1)
-        wref = weakref.ref(e1)
-        del e1, e2, e3
-        gc_collect()
-        self.assertIsNone(wref())
+        with self.subTest("Test the shortest cycle: d->element->d"):
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:2107]: Unrelated changes.

File: Lib/xml/etree/ElementTree.py [hunk 13]
```diff
@@ -196,12 +197,35 @@ def copy(self):
...
+        return self.__copy__()
+
+    def __copy__(self):
         elem = self.makeelement(self.tag, self.attrib)
         elem.text = self.text
         elem.tail = self.tail
         elem[:] = self
         return elem
 
+    def __deepcopy__(self, memo):
```
Comment (serhiy-storchaka) [inline Lib/xml/etree/ElementTree.py:211]: This method is not needed in Python implementation.

File: Modules/_elementtree.c [hunk 14]
```diff
@@ -283,10 +283,12 @@ create_new_element(PyObject* tag, PyObject* attrib)
     PyObject_GC_Track(self);
 
     if (attrib != Py_None && !is_empty_dict(attrib)) {
+        attrib = PyDict_Copy(attrib);
```
Comment (serhiy-storchaka) [inline Modules/_elementtree.c:298]: Why this change has been added?

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
`__deepcopy__` is not needed in Python implementation. Many tests duplicate already existing tests or expose implementation details.

File: Lib/xml/etree/ElementTree.py (hunk 13 above)
Comment (GPHemsley) [reply Lib/xml/etree/ElementTree.py:211]: I would prefer you have this discussion on the bpo ticket, as PEP 399 suggests otherwise.

File: Modules/_elementtree.c (hunk 14 above)
Comment (GPHemsley) [reply Modules/_elementtree.c:298]: Because the results of the tests in this commit uncovered a discrepancy between the C and Python implementations whereby the attrib list is not copied in the C version beyond the `__init__` function. (The Python implementation always uses `__init__` whenever it creates a new element, but the C implementation uses `create_new_element` in all locations except the original `__init__` function.)

File: Lib/test/test_xml_etree.py (hunk 12 above)
Comment (GPHemsley) [reply Lib/test/test_xml_etree.py:2107]: These changes helped me to debug the changes I made to the C implementation.

File: Lib/test/test_xml_etree.py (hunk 8 above)
Comment (GPHemsley) [reply Lib/test/test_xml_etree.py:148]: Shouldn't that be definitively decided one way or the other?

**Comment by GPHemsley** [comment]
Generally: I have added the unit tests in the way that I have because the existing checks are generally integration tests, and most of the test coverage for this module comes as a side effect of `pickle` tests. It is my goal to expand the unit tests later to test each method individually, so starting that work here makes it easier.

File: Lib/test/test_xml_etree.py (hunk 8 above)
Comment (serhiy-storchaka) [reply Lib/test/test_xml_etree.py:148]: Actually it already is tested in test_set_attribute.

File: Lib/test/test_xml_etree.py (hunk 12 above)
Comment (serhiy-storchaka) [reply Lib/test/test_xml_etree.py:2107]: Then please remove them after finishing debugging.

File: Lib/xml/etree/ElementTree.py (hunk 13 above)
Comment (serhiy-storchaka) [reply Lib/xml/etree/ElementTree.py:211]: This doesn't contradict PEP 399.

File: Modules/_elementtree.c (hunk 14 above)
Comment (serhiy-storchaka) [reply Modules/_elementtree.c:298]: In many case this makes a redundant copying since `create_new_element` is called after copying a dict.

File: Lib/test/test_xml_etree.py (hunk 12 above)
Comment (GPHemsley) [reply Lib/test/test_xml_etree.py:2107]: My point was that they were not wholly unrelated. What is the harm in including them here?

File: Lib/test/test_xml_etree.py (hunk 8 above)
Comment (scoder) [reply Lib/test/test_xml_etree.py:148]: It's not an implementation detail and there is a general difference between "no text" (i.e. None) and "empty text" (i.e. the empty string), which can be used for formatting.

Agreed that we should avoid duplicating tests, though. Preferably, there should be a single "obvious" place where a feature or property is tested.

File: Lib/test/test_xml_etree.py (hunk 12 above)
Comment (scoder) [reply Lib/test/test_xml_etree.py:2107]: I agree with @GPHemsley that it's better to have subtests than multiple tests in a single test function. Although having multiple test functions also wouldn't be all too bad. :)
Yes, they are unrelated to this specific ticket, but they don't really seem big enough to merit their own ticket and PR. They somewhat match the general refactoring of lengthy tests into subtests that this change does in other places. It's at least an edge case.

File: Lib/xml/etree/ElementTree.py (hunk 13 above)
Comment (scoder) [reply Lib/xml/etree/ElementTree.py:211]: The intention of PEP-399 is that both implementations *behave* the same. It's not that they use exactly the same mechanisms to reach that goal. Some things work in Python that don't easily work in C, and vice-versa. That's why there *are* two implementations in the first place.

If we can avoid adding Python (or C) code without sacrificing that goal, then that is a good thing. Code that does not exist is entirely free of bugs, for example. :)

File: Modules/_elementtree.c (hunk 14 above)
Comment (scoder) [reply Modules/_elementtree.c:298]: Can this be moved out of this function into the caller(s) that need it?

File: Modules/_elementtree.c [hunk 15]
```diff
@@ -725,6 +727,24 @@ _elementtree_Element___copy___impl(ElementObject *self)
...
+/*[clinic input]
+_elementtree.Element.copy
+
+[clinic start generated code]*/
+
+static PyObject *
+_elementtree_Element_copy_impl(ElementObject *self)
+/*[clinic end generated code: output=84660e8524276b22 input=1f8134305a7719a3]*/
+{
+    if (PyErr_WarnEx(PyExc_DeprecationWarning,
```
Comment (scoder) [inline Modules/_elementtree.c:773]: I'm personally ok with deprecating this right away (lxml never provided such a `.copy()` method, AFAICT), but still, the usual process would be to issue a `PendingDeprecationWarning` first.

**Review by scoder — CHANGES_REQUESTED** [review]
I agree with @serhiy-storchaka that some cleanup is needed to get this merged.

**Comment by GPHemsley** [comment]
@scoder [My latest comment on bpo-32424](https://bugs.python.org/issue32424#msg310362) explains some of my decisions. Would you mind providing feedback on that?

(In the meantime, I've rebased this against master.)

**Comment by GPHemsley** [comment]
Upon further review, it's clear I'm trying to do too many (4) different things in this one commit, so I'm going to break it up and submit the changes separately.
